plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/preferences/RunOptionsDefaultsComponent.java

```diff
@@ -153,35 +151,134 @@ public RunOptionsDefaultsComponent(Composite target, int columns, MessageTarget
+    stagingLocationInput.addModifyListener(new ModifyListener() {
+      @Override
+      public void modifyText(ModifyEvent e) {
+        validate();
+      }
+    });
 
-    updateStagingLocations(project);
+    updateStagingLocations(project, 0); // no delay
     messageTarget.setInfo(Messages.getString("set.pipeline.run.option.defaults")); //$NON-NLS-1$
+    validate();
+  }
+
+  private void validate() {
+    setPageComplete(false);
+    Credential selectedCredential = accountSelector.getSelectedCredential();
+    if (selectedCredential == null) {
+      projectInput.setEnabled(false);
+      stagingLocationInput.setEnabled(false);
+      createButton.setEnabled(false);
+      return;
+    }
+
+    projectInput.setEnabled(true);
+    if (Strings.isNullOrEmpty(projectInput.getText())) {
+      stagingLocationInput.setEnabled(false);
+      createButton.setEnabled(false);
+      return;
+    }
+    // FIXME: incorporate project verification here
+
+    stagingLocationInput.setEnabled(true);
+
+    final String bucketNamePart = GcsDataflowProjectClient.toGcsBucketName(getStagingLocation());
+    if (bucketNamePart.isEmpty()) {
+      // If the bucket name is empty, we don't have anything to verify; and we don't have any
+      // interesting messaging.
+      createButton.setEnabled(false);
+      return;
+    }
+
+    IStatus status = bucketNameValidator.validate(bucketNamePart);
+    if (!status.isOK()) {
+      messageTarget.setError(status.getMessage());
+      createButton.setEnabled(false);
+      return;
+    }
+    
+    // we have a project and staging location, so a fetch-staging-locations job should be under way
+    Future<SortedSet<String>> stagingLocationsFuture = getStagingLocationsResult();
+    Preconditions.checkNotNull(stagingLocationsFuture, "Fetch staging locations job should be underway");
+    if (stagingLocationsFuture.isDone()) {
+      try {
+        // on error, will raise an exception
+        stagingLocationsFuture.get();
+      } catch (ExecutionException ex) {
+        messageTarget.setError(Messages.getString("could.not.retrieve.buckets.for.project", //$NON-NLS-1$
+            projectInput.getText()));
+        DataflowUiPlugin.logError(ex, "Exception while retrieving staging locations"); //$NON-NLS-1$
+        return;
+      } catch (InterruptedException ex) {
+        DataflowUiPlugin.logError(ex, "Interrupted while retrieving staging locations"); //$NON-NLS-1$
+      }
+    }
+
+    // we have a project and staging location, so a verify job should be under way
+    Future<VerifyStagingLocationResult> future = getVerifyStagingLocationResult();
+    Preconditions.checkNotNull(future, "verification job should be underway");
+    if (!future.isDone()) {
+      messageTarget.setInfo("Verifying staging location...");
+      createButton.setEnabled(false);
+      return;
+    }
+
+    try {
+      VerifyStagingLocationResult result = future.get();
+      if (!result.email.equals(accountSelector.getSelectedEmail())
```

Comment:
Previously, `VerifyStagingLocationResult.email` was needed to get the email given to the job for this check at this point. Now that the field `verifyStagingLocationJob` has the `getEmail()` field, I wonder if we want to remove `.email` from the result type. (The result is from the current `verifyStagingLocationJob`.) However, it may still be harmless to keep `.email`.

The same argument applies to the `.stagingLocation` field.